File: solutions/erlang/square-root/1/src/square_root.erl
-module(square_root).

-export([square_root/1]).

square_root(Radicand) -> square_root(Radicand, 1).

square_root(Radicand, Candidate) when Candidate * Candidate == Radicand -> Candidate;
square_root(Radicand, Candidate) when Candidate * Candidate < Radicand -> square_root(Radicand, Candidate + 1).

File: solutions/elixir/square-root/1/lib/square_root.ex
defmodule SquareRoot do
  @doc """
  Calculate the integer square root of a positive integer
  """
  @spec calculate(radicand :: pos_integer) :: pos_integer
  def calculate(radicand) do
    sqrt(radicand)
  end

  defp sqrt(n, i \\ 1) do
    case n do
      n when n == i*i -> i
      n when n < i*i -> -1
      _ -> sqrt(n, i+1)
    end
  end
end
